### ISSUE TITLE
Custom syntax to allow highlighting of specific words

### ISSUE THREAD

**Issue by mYnDstrEAm** [jaap-karssenberg/zim-desktop-wiki#186]
It would be great if Zim allowed to highlight specific words (and/or whole lines after specific words) similar to how [http://docs.notepad-plus-plus.org/index.php/User_Defined_Languages Notepad++ allows this via custom languages].

**Comment by jaap-karssenberg** [comment]
What would be your use case for that ?

**Comment by mystyc** [comment]
It works sort of like an instant search and resembles Zim's find & highlight, except it has 3 additional important features:
1. Notepad++ gives 5 "slots" for this feature which can highlight 5 different words with 5 different colors in the same document.  
2. A shortcut key can be set that moves the cursor to each of the other occurrences of this word in a document.
3. It is separate from the search/find feature, thus freeing the user to use search/find on a 6th word or some sort of more sophisticated search query.

### Example Usages
The most obvious use would be that it effectively allows for multiple search queries to run simultaneously.  
More specifically, let's say there's a document of python code, and you come across `Object_Instance.Method(parameter)`, and now you want to follow every occurrence of the "_Object_Instance_," "_Method_," and "_parameter_," even though "_Object_Instance.Method(parameter)_" occurs only once in the document.  For me in this situation, I highlight each word separately and use the shortcut key combo `CTRL+NumPad1`, `CTRL+NumPad2`, and `CTRL+NumPad3` respectively for each word; each highlighting their word with either blue, green, or yellow, respectively.  Then I use the shortcut `CTRL+1`, `CTRL+2`, and `CTRL+3` respectively to cycle through each word separately.

**Comment by jaap-karssenberg** [comment]
I see how this could useful for a code editor, but afraid I fail to see the usefulness for a note application like zim

**Comment by mystyc** [comment]
Okay.  Well, are you familiar with the practice of using physical highlighter markers to highlight passages in a textbook or notebook?  In those situation, the markers are used to highlight certain keywords, sentences, phrases, or even entire paragraphs as part of an effort to study for a test or do some quick research.  Those who often use this method will tend to have markers of several different colors that are each designated for a certain word or concept.  The idea is that it helps in distinguishing important content from the less important stuff.  

As a more specific example, I once used this feature while I was reading an article about a famous married couple who are each individually famous for different reasons.  I was only interested in one of them, and I was especially interested in copying direct and indirect quotes.  The article had both direct and indirect quotes from each of them, but the author had an annoying tendency to use the last name to refer to either and/or both of them.

After copying the text into Notepad++, I assigned color-1 to spouse-1, color-2 to spouse-2, and color-3 to their shared last name.  Then I used color-4 on quotation marks and on the words "said" and "told." 
This wasn't preplanned, rather, I created this schema as I was reading.
  As a result, I was able to pull some indirect quotes from spouse-1 without reading the whole article.

The text copy was actually stored in Zim because Zim makes it easy to search through my entire collection at once, while also me to do something similar with my own original notes.  Frequently I find myself copying short text documents from Zim into Notepad++ in order to use this feature.  

I'd probably use this feature more if it was in Zim, and it would likely change the way I take notes in ways that I haven't thought of yet.

In Notepad++, the above feature only works within a single open document, even though N++ is capable of searching multiple documents at once.  If this multi-highlighting feature was also part of Zim's "global search" function, then I would be able to do the above on a more massive scale.

**Comment by jaap-karssenberg** [comment]
The multiple word search still sounds rather complicated to me.

Maybe an easier version would be to have a button that formats all search results with "mark". You won't get different colors (zim wiki syntax doesn't support that) but you can do multiple searches and have the results remain highlighted.

Would that work?

Sounds like a small plugin someone could write that adds an action.

**Comment by mystyc** [comment]
That would just be search and replace with refreshing the page.
_Find_ `Word`, _Replace with_  `__Word__`, _refresh_.
That's a permanent change which is entirely different.  
The current search function is capable of highlighting text in a color that is different that which is used for the standard text formats (strong/emphasis/verbatim...).  The effect goes away once the search box is closed.  The feature above is like that, except it does not have a search box, but even if it worked with the search box, that would have a similar effect.  
Does that search and highlight feature use "wiki syntax?"  
I thought it was perhaps part of GTK.

**Comment by jaap-karssenberg** [comment]
Hmm, from your description above I got the strong impression that you were
looking for permanent highlighting. Bit lost what you mean by "custom
syntax" in the title now.  The notepad+ documentation you refer to is about
syntax highlighting, zim does not support this (except in code blocks), but
it sounds to me this is a solution to an underlying use case, not the
essence of the request.

Search highlighting is indeed done on top of the user interface, not in the
wiki source.

So are you just looking for search with multiple terms?  This could be a
straight-forward extension borrowing regular expression syntax, e.g.
"name1|name2" could match two names and highlight them.

Different color coding per regex group is not that hard either.

-- Jaap

**Comment by mystyc** [comment]
Oh, sorry, I totally forgot OP of this thread.  
For some reason I read this as the N++ feature I described above, but I see now that @mYnDstrEAm was referring to N++'s built-in Scintilla lexer and style configuration, as described here, [docs.notepad-plus-plus.org/index.php/Built-in_Languages](http://docs.notepad-plus-plus.org/index.php/Built-in_Languages).

In Zim, the lexer more closely resembles the GTK code block languages.  Basically, N++ can do that for an entire document instead of just a window, though on the flip-side, Zim can use multiple "lexical styles" on the same page via different code blocks.

I agree that implementing this, or even expanding this aspect of the code block would probably be too much for Zim.  I think it would basically be equivalent allowing the user to arbitrarily switch out the wiki syntax with any of a list of other syntactical styles.  

We can probably close this specific issue and move our discussion to a new feature request issue with a title like, "_Allow temporary highlighting of specific words for several different words/colors concurrently_," or something (I don't know how to do that).
For reference, this N++ feature is described fairly well at, [NotePad++ Marking And Highlighting](http://docs.notepad-plus-plus.org/index.php/Marking_And_Highlighting).

**Comment by jaap-karssenberg** [comment]
CLosing this item as Won't Fix. Please open a new issue if you have a concrete use case for zim in particular.
